backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt

```diff
@@ -1397,27 +1398,33 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
             val thisPtr = evaluateExpression(value.receiver!!)
             return functionGenerationContext.loadSlot(
                     fieldPtrOfClass(thisPtr, value.symbol.owner), value.descriptor.isVar())
-        }
-        else {
+        } else {
             assert (value.receiver == null)
             val ptr = context.llvmDeclarations.forStaticField(value.symbol.owner).storage
             return functionGenerationContext.loadSlot(ptr, value.descriptor.isVar())
         }
     }
 
     //-------------------------------------------------------------------------//
+    private fun needMutationCheck(expression: IrExpression): Boolean {
+        // For now we omit mutation checks on immutable types, as this allows initialization in constructor
+        // and it is assumed API doesn't allow to change them.
+        val descriptor = TypeUtils.getClassDescriptor(expression.type)
+        return descriptor?.annotations?.findAnnotation(FqName("konan.Immutable")) == null
+    }
 
     private fun evaluateSetField(value: IrSetField): LLVMValueRef {
         context.log{"evaluateSetField               : ${ir2string(value)}"}
         val valueToAssign = evaluateExpression(value.value)
         if (value.descriptor.dispatchReceiverParameter != null) {
             val thisPtr = evaluateExpression(value.receiver!!)
-            functionGenerationContext.call(context.llvm.mutationCheck,
-                    listOf(functionGenerationContext.bitcast(codegen.kObjHeaderPtr, thisPtr)),
-                    Lifetime.IRRELEVANT, ExceptionHandler.Caller)
+            if (needMutationCheck(value.receiver!!)) {
```

Comment:
The check based on receiver type doesn't look reliable to me. We can get `ClassDescriptor` from field descriptor instead, e.g.`value.descriptor.containingDeclaration`.

Reply:
It does not have `findAnnotation`, it seems.